[DataPipe] Update `mux` data pipe (#76384)

ninginthecloud opened the pull request:
Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76384

OSS issue discussion: https://github.com/pytorch/data/issues/346
This diff updates `mux` and `mux_longest` data pipe.
`mux`: Yields one element at a time from each of the input Iterable DataPipes (functional name: ``mux``). As in, one element from the 1st input DataPipe, then one element from the 2nd DataPipe in the next iteration, and so on. It ends when the shortest input DataPipe is exhausted.

`mux` example:

```
>>> from torchdata.datapipes.iter import IterableWrapper
>>> dp1, dp2, dp3 = IterableWrapper(range(3)), IterableWrapper(range(10, 15)), IterableWrapper(range(20, 25))
>>> list(dp1.mux(dp2, dp3))
[0, 10, 20, 1, 11, 21, 2, 12, 22]
```

Test Plan:
buck test mode/opt //caffe2/test:datapipe

https://www.internalfb.com/intern/testinfra/testrun/4785074706282345

Differential Revision: D36017945

## BC-breaking Note:
- The modified `mux` datapipe will have different expected result compared to previous one. 
- After this PR, `mux` datapipe yields one element at a time from each of the iterable DataPipe input. As in, one element from the 1st input DataPipe, then one element from the 2nd DataPipe in the next iteration, and so on. It ends when the **shortest** input DataPipe is exhausted.
- `len(dp)` will be the multiplication of length of shortest Datapipe and the number of datapipes. While, previosly, `len(dp)` is the sum of all datapipe's length. 
For example,

```
from torchdata.datapipes.iter import IterableWrapper
dp1, dp2, dp3 = IterableWrapper(range(3)), IterableWrapper(range(10, 15)), IterableWrapper(range(20, 25))
output_dp = dp1.mux(dp2, dp3)
```

### 1. 11
`output_dp` ends when all intput DataPipe has exhausted. 
```
>>> list(output_dp)
[0, 10, 20, 1, 11, 21, 2, 12, 22, 3, 13, 23, 4, 14, 24]
>>> len(output_dp)
13
```

### This PR:
`output_dp` ends when the shorted datapipe has exhausted. 
```
>>> list(output_dp)
[0, 10, 20, 1, 11, 21, 2, 12, 22]
>>> len(output_dp)
9
```

For users who to combine all iterable datapipe inputs but ends when all input DataPipe is exhausted, they can use [`mux_longest`](https://github.com/pytorch/data/blob/0e6554a5cb2747f01e39dc529e21009ff12e4e7a/torchdata/datapipes/iter/util/mux_longest.py),  introduced by PR https://github.com/pytorch/data/pull/372 in torchdata. 
For example, 
```
>>> from torchdata.datapipes.iter import IterableWrapper
>>> dp1, dp2, dp3 = IterableWrapper(range(3)), IterableWrapper(range(10, 15)), IterableWrapper(range(20, 25))
>>> list(dp1.mux_longest(dp2, dp3))
[0, 10, 20, 1, 11, 21, 2, 12, 22, 3, 13, 23, 4, 14, 24]
```

NivekT reviewed (approved):
LGTM! Just one small comment about buffer, please see my comment in the internal diff.

NivekT reviewed (changes requested):
FYI, there is current one CI test failing:

```
Traceback (most recent call last):
  File "test_datapipe.py", line 2247, in test_simple_sharding
    self.assertEqual([1, 20, 40, 70], items)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 2258, in assertEqual
    msg=msg,
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_comparison.py", line 1049, in assert_equal
    raise error_meta.to_error() from None
AssertionError: The length of the sequences mismatch: 4 != 2
```

NivekT reviewed (approved):
CIs look good so far. We will wait a little bit til it is done to confirm, but LGTM!

Thanks for submitting the PR!

ninginthecloud committed:
[DataPipe] Update `mux` data pipe (#77145)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/77145

Pull Request resolved: https://github.com/pytorch/pytorch/pull/76384

OSS issue discussion: https://github.com/pytorch/data/issues/346
This diff updates `mux` and `mux_longest` data pipe.
`mux`: Yields one element at a time from each of the input Iterable DataPipes (functional name: ``mux``). As in, one element from the 1st input DataPipe, then one element from the 2nd DataPipe in the next iteration, and so on. It ends when the shortest input DataPipe is exhausted.

`mux` example:

```
>>> from torchdata.datapipes.iter import IterableWrapper
>>> dp1, dp2, dp3 = IterableWrapper(range(3)), IterableWrapper(range(10, 15)), IterableWrapper(range(20, 25))
>>> list(dp1.mux(dp2, dp3))
[0, 10, 20, 1, 11, 21, 2, 12, 22]
```

Test Plan:
buck test mode/opt //caffe2/test:datapipe

https://www.internalfb.com/intern/testinfra/testrun/5910974614015948

Reviewed By: NivekT, ejguan

Differential Revision: D36017945

fbshipit-source-id: b35fb0d3f9df3610e663790b7851c8f40d20f594
